In [2]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [3]:
# Load the given datasets
kidneyDiseaseNumerical = pd.read_csv("chronic_kidney_disease_numerical.csv")
kidneyDiseaseCategorical = pd.read_csv("chronic_kidney_disease_categorical.csv")

# Print the data
print("Categorical Kidney Disease Data:")
print(kidneyDiseaseCategorical)
print("\nNumerical Kidney Disease Data:")
print(kidneyDiseaseNumerical)

Categorical Kidney Disease Data:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0 

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The data is a collection of information from a patient about specific medical categories and quantities. The given datasets contain both categorial and numerical data. Three examples of categorical data include specific gravity (sg), albumin (al), and sugar (su). Three examples of numerical data include age, blood pressure (bp), and blood glucose random (bgr).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [4]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicateNumerical = kidneyDiseaseNumerical.duplicated()
totalDuplicateNumerical = duplicateNumerical.sum()
print("Total duplicate rows in numerical dataset: ", totalDuplicateNumerical)

# Drop these duplicate rows
if totalDuplicateNumerical > 0:
    kidneyDiseaseNumerical = kidneyDiseaseNumerical[~duplicateNumerical]

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicateCategorical = kidneyDiseaseCategorical.duplicated()
totalDuplicateCategorial = duplicateCategorical.sum()
print("Total duplicate rows in categorical dataset: ", totalDuplicateCategorial)

if totalDuplicateCategorial > 0:
    kidneyDiseaseCategorical = kidneyDiseaseCategorical[~duplicateCategorical]


Total duplicate rows in numerical dataset:  30
Total duplicate rows in categorical dataset:  30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: I don't think all of the uniue ids are present in both datasets because in Step 1 we saw that the amount of rows in the datasets were not the same and when we removed duplicates there were the same amount of duplicates meaning that one of the datasets still has more rows than the other. This means one dataset must have more unique ids than the other. Since the categorical dataset is the left dataset, the rows were included and where they did not merge with the numerical dataset, the entries in those columns become NaN. 

In [5]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
mergedDataset = pd.merge(kidneyDiseaseCategorical, kidneyDiseaseNumerical, on='unique_id')

#Print the combined dataset
print("Combined dataset:")
print(mergedDataset)

Combined dataset:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
385     789906  0.0  0.0    normal    normal  notpresent  notpresent   no   
386     490253  0.0  0.0    normal    normal  notpresent  notpresent   no   
387     492993  3.0  0.0       NaN    normal  notpresent  notpresent  yes   
388     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   
389     583971  1.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm cad  ...    bp    bgr     bu   sc    sod  pot  h

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [6]:
# Calculate the percentage of rows that contain atleast one missing value
percentageMissing = mergedDataset.isnull().any(axis=1).mean() * 100

# Print %
print("Percentage of rows with at least one missing value: {:.2f}%".format(percentageMissing))

# Drop these rows from the dataset
mergedDatasetDropped = mergedDataset.dropna()

# Print the Dataset
print("\nDataset after dropping rows with missing values:")
print(mergedDatasetDropped)

Percentage of rows with at least one missing value: 60.77%

Dataset after dropping rows with missing values:
     unique_id   al   su       rbc      pc         pcc          ba  htn   dm  \
4       881763  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
5       114717  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
7       901528  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
8       433613  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
9       621332  1.0  0.0  abnormal  normal  notpresent  notpresent   no  yes   
..         ...  ...  ...       ...     ...         ...         ...  ...  ...   
382     996076  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
384     397388  2.0  2.0    normal  normal  notpresent     present  yes   no   
385     789906  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
386     490253  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
388     273

## Part 3.4: Sort the dataset according to the Labels

In [7]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sortedDataset = mergedDatasetDropped.sort_values(by='Target')
sortedDataset.reset_index(drop=True, inplace=True)

# Print the dataset
print("\nSorted dataset according to 'Target' column:")
print(sortedDataset)


Sorted dataset according to 'Target' column:
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       534245  2.0  0.0    normal  abnormal     present  notpresent   no   
1       349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
2       343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
4       514721  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
148     489160  0.0  0.0    normal    normal  notpresent  notpresent   no   
149     530138  0.0  0.0    normal    normal  notpresent  notpresent   no   
150     261137  0.0  0.0    normal    normal  notpresent  notpresent   no   
151     322882  0.0  0.0    normal    normal  notpresent  notpresent   no   
152     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...    bp    b

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [8]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dummyDataset = pd.get_dummies(sortedDataset, columns=categorical_columns)

# Print the dataset
print("\nDataset after converting categorical columns to dummy variables:")
print(dummyDataset)


Dataset after converting categorical columns to dummy variables:
     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0       534245  2.0  0.0  45.0  70.0  117.0   52.0   2.2  136.0  3.8  ...   
1       349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  ...   
2       343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2  ...   
3       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
4       514721  2.0  0.0  56.0  90.0  129.0  107.0   6.7  131.0  4.8  ...   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...  ...   
148     489160  0.0  0.0  28.0  60.0   79.0   50.0   0.5  145.0  5.0  ...   
149     530138  0.0  0.0  23.0  80.0  111.0   34.0   1.1  145.0  4.0  ...   
150     261137  0.0  0.0  35.0  60.0  105.0   39.0   0.5  135.0  3.9  ...   
151     322882  0.0  0.0  33.0  80.0  128.0   38.0   0.6  135.0  3.9  ...   
152     273038  0.0  0.0  47.0  60.0  117.0   22.0   1.2  138.0  3.5  ...   

     cad_

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [9]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for column in numerical_columns:
    mean = dummyDataset[column].mean()
    stdDev = dummyDataset[column].std()

    lowerBound = mean - 3 * stdDev
    upperBound = mean + 3 * stdDev

    dummyDataset = dummyDataset[(dummyDataset[column] >= lowerBound) & (dummyDataset[column] <= upperBound)]

#Reindex the dataset
dummyDataset.reset_index(drop=True, inplace=True)

# Print the dataset
print("\nDataset after removing outlier from numerical columns:")
print(dummyDataset)


Dataset after removing outlier from numerical columns:
     unique_id   al   su   age    bp    bgr     bu   sc    sod  pot  ...  \
0       343710  2.0  0.0  61.0  80.0  173.0  148.0  3.9  135.0  5.2  ...   
1       514721  2.0  0.0  56.0  90.0  129.0  107.0  6.7  131.0  4.8  ...   
2       124304  3.0  4.0  69.0  70.0  264.0   87.0  2.7  130.0  4.0  ...   
3       980291  3.0  0.0  62.0  70.0  122.0   42.0  1.7  136.0  4.7  ...   
4       615697  4.0  0.0   6.0  60.0   94.0   67.0  1.0  135.0  4.9  ...   
..         ...  ...  ...   ...   ...    ...    ...  ...    ...  ...  ...   
128     489160  0.0  0.0  28.0  60.0   79.0   50.0  0.5  145.0  5.0  ...   
129     530138  0.0  0.0  23.0  80.0  111.0   34.0  1.1  145.0  4.0  ...   
130     261137  0.0  0.0  35.0  60.0  105.0   39.0  0.5  135.0  3.9  ...   
131     322882  0.0  0.0  33.0  80.0  128.0   38.0  0.6  135.0  3.9  ...   
132     273038  0.0  0.0  47.0  60.0  117.0   22.0  1.2  138.0  3.5  ...   

     cad_no  cad_yes  appet_goo

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [10]:
# Normalize the all Numerical Attributes in the dataset.
for column in numerical_columns:
    mean = dummyDataset[column].mean()
    stdDev = dummyDataset[column].std()

    dummyDataset[column] = (dummyDataset[column] - mean) / stdDev

# Print the dataset
print("\nDataset after normalizing numerical columns:")
print(dummyDataset)


Dataset after normalizing numerical columns:
     unique_id   al   su       age        bp       bgr        bu        sc  \
0       343710  2.0  0.0  0.807138  0.727802  1.142120  3.999185  1.683685   
1       514721  2.0  0.0  0.492884  1.746724  0.169523  2.466464  3.564322   
2       124304  3.0  4.0  1.309946 -0.291121  3.153626  1.718795  0.877698   
3       980291  3.0  0.0  0.869989 -0.291121  0.014792  0.036540  0.206042   
4       615697  4.0  0.0 -2.649663 -1.310043 -0.604133  0.971126 -0.264117   
..         ...  ...  ...       ...       ...       ...       ...       ...   
128     489160  0.0  0.0 -1.266943 -1.310043 -0.935700  0.335608 -0.599945   
129     530138  0.0  0.0 -1.581197  0.727802 -0.228357 -0.262527 -0.196952   
130     261137  0.0  0.0 -0.826986 -1.310043 -0.360984 -0.075610 -0.599945   
131     322882  0.0  0.0 -0.952688  0.727802  0.147419 -0.112994 -0.532780   
132     273038  0.0  0.0 -0.072775 -1.310043 -0.095731 -0.711129 -0.129786   

          sod    

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Unique ID
- Reason: Unique IDs will not have any useful information on deciding which factors go along with CKD, they are simply identifiers and not any medical or lab results.


In [11]:
# Remove that column
columns_to_remove = ['unique_id']
dummyDataset.drop(columns=columns_to_remove, inplace=True)

# Print the dataset
print("\nDataset after removing less informative columns:")
print(dummyDataset)


Dataset after removing less informative columns:
      al   su       age        bp       bgr        bu        sc       sod  \
0    2.0  0.0  0.807138  0.727802  1.142120  3.999185  1.683685 -0.956087   
1    2.0  0.0  0.492884  1.746724  0.169523  2.466464  3.564322 -1.657657   
2    3.0  4.0  1.309946 -0.291121  3.153626  1.718795  0.877698 -1.833050   
3    3.0  0.0  0.869989 -0.291121  0.014792  0.036540  0.206042 -0.780695   
4    4.0  0.0 -2.649663 -1.310043 -0.604133  0.971126 -0.264117 -0.956087   
..   ...  ...       ...       ...       ...       ...       ...       ...   
128  0.0  0.0 -1.266943 -1.310043 -0.935700  0.335608 -0.599945  0.797838   
129  0.0  0.0 -1.581197  0.727802 -0.228357 -0.262527 -0.196952  0.797838   
130  0.0  0.0 -0.826986 -1.310043 -0.360984 -0.075610 -0.599945 -0.956087   
131  0.0  0.0 -0.952688  0.727802  0.147419 -0.112994 -0.532780 -0.956087   
132  0.0  0.0 -0.072775 -1.310043 -0.095731 -0.711129 -0.129786 -0.429910   

          pot      hemo  

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [12]:
# Export the dataframe to a new csv file
dummyDataset.to_csv('ckd_results.csv')

# Export the dataframe to a new json file
dummyDataset.to_json('ckd_results.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [13]:
pd.set_option('display.max_columns', None)
top_15_rows = dummyDataset.head(15)

print("Top 15 rows of the cleaned dataset:")
print(top_15_rows)

Top 15 rows of the cleaned dataset:
     al   su       age        bp       bgr        bu        sc       sod  \
0   2.0  0.0  0.807138  0.727802  1.142120  3.999185  1.683685 -0.956087   
1   2.0  0.0  0.492884  1.746724  0.169523  2.466464  3.564322 -1.657657   
2   3.0  4.0  1.309946 -0.291121  3.153626  1.718795  0.877698 -1.833050   
3   3.0  0.0  0.869989 -0.291121  0.014792  0.036540  0.206042 -0.780695   
4   4.0  0.0 -2.649663 -1.310043 -0.604133  0.971126 -0.264117 -0.956087   
5   3.0  1.0  0.430033  0.727802  2.048403  1.195427  1.683685 -0.605302   
6   1.0  0.0 -0.135626 -1.310043  0.921075  1.905712  1.280691  0.096268   
7   4.0  2.0  0.995691  2.765646  0.921075  0.485141  3.900150 -0.079124   
8   4.0  0.0 -0.009924  0.727802 -0.582029  4.559936  4.235978 -0.780695   
9   4.0  3.0  1.309946 -0.291121  2.048403  2.055246  3.295659 -3.586975   
10  3.0  2.0  1.561350  2.765646  3.838865  1.830945  2.825500 -0.079124   
11  4.0  0.0 -1.706899  1.746724 -0.316775 -0.038227

### Paste the markdown table here

|   al | su  |     age    |     bp    |     bgr   |     bu    |    sc    |     sod    |    pot    |    hemo   |    pcv   |   wbcc  |   rbcc  | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|------|-----|------------|-----------|-----------|-----------|----------|------------|-----------|-----------|----------|---------|---------|--------------|------------|-------------|-----------|----------------|-------------|----------------|------------|--------|---------|-------|--------|--------|---------|------------|------------|-------|--------|--------|---------|------------|---------------|
|  2.0 | 0.0 | 0.807138   | 0.727802  | 1.142120  | 3.999185  | 1.683685 | -0.956087  | 1.367853  | -2.902248 | -2.894560 | 0.517701 | -2.293292 | True         | False      | True        | False     | True           | False       | True           | False      | False  | True    | False | True   | False  | True    | False      | True       | False | True   | False  | True    | True       | False         |
|  2.0 | 0.0 | 0.492884   | 1.746724  | 0.169523  | 2.466464  | 3.564322 | -1.657657  | 0.740527  | -2.295400 | -2.173900 | -0.718144 | -2.050645 | True         | False      | True        | False     | True           | False       | True           | False      | False  | True    | True  | False  | True   | False   | True       | False      | True  | False  | True   | False   | True       | False         |
|  3.0 | 4.0 | 1.309946   | -0.291121 | 3.153626  | 1.718795  | 0.877698 | -1.833050  | -0.514124 | -0.821625 | -1.020844 | 0.694250  | -1.201378 | False        | True       | True        | False     | True           | False       | True           | False      | False  | True    | False | True   | False  | True    | True       | False      | False | True   | True   | False   | True       | False         |
|  3.0 | 0.0 | 0.869989   | -0.291121 | 0.014792  | 0.036540  | 0.206042 | -0.780695  | 0.583696  | -0.778278 | -0.732580 | -0.056084 | -1.444026 | False        | True       | True        | False     | True           | False       | True           | False      | False  | True    | False | True   | True   | False   | True       | False      | True  | False  | True   | False   | True       | False         |
|  4.0 | 0.0 | -2.649663  | -1.310043 | -0.604133 | 0.971126  | -0.264117| -0.956087  | 0.897359  | -1.948629 | -2.029768 | 3.827999  | -0.352112 | True         | False      | True        | False     | True           | False       | False          | True       | True   | False   | True  | False  | True   | False   | False      | True       | True  | False  | True   | False   | True       | False         |
|  3.0 | 1.0 | 0.430033   | 0.727802  | 2.048403  | 1.195427  | 1.683685 | -0.605302  | 0.897359  | -1.515166 | -1.453240 | -0.276771 | -1.686673 | False        | True       | True        | False     | False          | True        | False          | True       | False  | True    | False | True   | True   | False   | True       | False      | False | True   | True   | False   | True       | False         |
|  1.0 | 0.0 | -0.135626  | -1.310043 | 0.921075  | 1.905712  | 1.280691 | 0.096268   | -0.514124 | -1.991975 | -2.318032 | 2.901115  | -2.293292 | False        | True       | False       | True      | True           | False       | True           | False      | False  | True    | False | True   | True   | False   | True       | False      | True  | False  | True   | False   | True       | False         |
|  4.0 | 2.0 | 0.995691   | 2.765646  | 0.921075  | 0.485141  | 3.900150 | -0.079124  | 0.426864  | -2.815556 | -2.606296 | -0.232633 | -2.050645 | True         | False      | True        | False     | True           | False       | False          | True       | False  | True    | False | True   | True   | False   | True       | False      | False | True   | True   | False   | True       | False         |
|  4.0 | 0.0 | -0.009924  | 0.727802  | -0.582029 | 4.559936  | 4.235978 | -0.780695  | -0.827787 | -1.991975 | -1.741504 | -0.497457 | -2.050645 | False        | True       | True        | False     | True           | False       | True           | False      | False  | True    | True  | False  | True   | False   | True       | False      | True  | False  | False  | True    | True       | False         |
|  4.0 | 3.0 | 1.309946   | -0.291121 | 2.048403  | 2.055246  | 3.295659 | -3.586975  | -0.670955 | -2.165361 | -2.318032 | 1.532859  | -2.171968 | False        | True       | True        | False     | False          | True        | False          | True       | False  | True    | False | True   | False  | True    | True       | False      | False | True   | False  | True    | True       | False         |
|  3.0 | 2.0 | 1.561350   | 2.765646  | 3.838865  | 1.830945  | 2.825500 | -0.079124  | -2.239269 | -2.252053 | -2.029768 | -0.453320 | -2.293292 | True         | False      | True        | False     | False          | True        | True           | False      | False  | True    | False | True   | False  | True    | False      | True       | True  | False  | True   | False   | True       | False         |
|  4.0 | 0.0 | -1.706899  | 1.746724  | -0.316775 | -0.038227 | 0.206042 | -2.710013  | -1.298281 | -2.642170 | -3.038692 | 1.930095  | -1.444026 | False        | True       | True        | False     | False          | True        | False          | True       | True   | False   | True  | False  | True   | False   | True       | False      | True  | False  | False  | True    | True       | False         |
|  4.0 | 1.0 | 0.995691   | -1.310043 | 2.601015  | 0.634675  | 1.952347 | -0.605302  | 1.681516  | -2.122014 | -2.173900 | -0.232633 | -2.050645 | True         | False      | True        | False     | True           | False       | False          | True       | False  | True    | False | True   | True   | False   | False      | True       | False | True   | True   | False   | True       | False         |
|  4.0 | 1.0 | 0.492884   | -0.291121 | 1.959985  | -0.561595 | 0.206042 | -0.780695  | -0.827787 | 0.738843  | 1.141135  | 1.974232  | 0.618477  | True         | False      | False       | True      | True           | False       | True           | False      | True   | False   | True  | False  | True   | False   | True       | False      | True  | False  | True   | False   | True       | False         |
|  4.0 | 0.0 | -0.009924  | 0.727802  | 0.257941  | 3.662734  | 4.773303 | -1.482265  | 1.838347  | -1.775244 | -1.164976 | -0.806419 | -1.322702 | True         | False      | True        | False     | True           | False       | False          | True       | True   | False   | False | True   | False  | True    | True       | False      | False | True   | True   | False   | True       | False         |



** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: I think there are two ways you can mitigate risks when it comes to data conversion, those being: learn prompt engineering in order to prompt the LLM to maximize the quality of the output; or simply do not use LLMs to handle the data at all. Prompt engineering can be useful for answering many questions when it comes to LLMs. If you ask them the right questions and give them all the information they need to make the best output, then your output will be better off. However, when it comes to large datasets, it is not helpful if the LLM is going to make up number or values and 'hallucinate'. Therefore, it would be best to use methods, like pandas, that have been tested and proven to work well with large sets of data, thus making them more useful and dependable when it comes to large sets of data.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [14]:
# Code to rename all the columns in the dataset
rename_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

sortedDataset.rename(columns=rename_mapping, inplace=True)

print("Result:")
print(sortedDataset.head(2))

Result:
   unique_id  albumin  sugar red blood cells  pus cell pus cell clumps  \
0     534245      2.0    0.0          normal  abnormal         present   
1     349892      1.0    3.0        abnormal  abnormal      notpresent   

     bacteria hypertension diabetes mellitus coronary artery disease appetite  \
0  notpresent           no                no                      no     good   
1  notpresent          yes               yes                     yes     good   

  pedal edema anemia Target   age  blood pressure  blood glucose random  \
0          no     no    ckd  45.0            70.0                 117.0   
1          no     no    ckd  59.0            70.0                 424.0   

   blood urea  serum creatinine  sodium  potassium  hemoglobin  \
0        52.0               2.2   136.0        3.8        10.0   
1        55.0               1.7   138.0        4.5        12.6   

   packed cell volume     wbcc  rbcc  
0                30.0  19100.0   3.7  
1                37.0 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [15]:
# Converted SQL to Pandas code
result = sortedDataset.groupby('Target').size().reset_index(name='count')
print("Result:")
print(result)

Result:
   Target  count
0     ckd     42
1  notckd    111


*Describe what the above code does here*

This code looked at the 'Target' column and saw that this column contained only two subsets: "ckd" or "notckd". It then counted the number of rows that had "ckd" as their entry in the 'Target' column. Then, it did the same for "notckd" by counting up the rows that the 'Target' column had "notckd" entered.